Event factory strategy

j-ka11 opened the pull request.

j-ka11 committed:
Created MeasurementEventFactory and MeasurementEvent classes.

The MeasurementEvent class provides an abstraction that contains all the information needed for the Injector class to inject javascript into the page to track an event for a plugin. The MeasurementEventFactory provides a function that the Injector class can use to produce a list of MeasurementEvent objects for a certain plugin.

j-ka11 committed:
Created MeasurementEventList class and WoocommerceEventList class.

MeasurementEventList is a template class that holds all the information needed to create a list of MeasurementEvent objects. It can be extended by a plugin-specific list to add its customized measurement event information. The WoocommerceEventList is an example of such a class and adds its woocommerce-specific information to create a list of trackable events.

j-ka11 committed:
Implemented Javascript injection to add event listeners to the Woocommerce events and log when one of those events occurs.

At this point, the actual css selectors for woocommerce need to be defined more clearly to make sure every event works for every theme. Another thing that is needed is to add javascript that checks if the element is actually on the page before we add an event listener to prevent the console from being overloaded with errors. On events that cause the page to reload, we are unable to see if the event fired so we need to come up with a solution for that as well. Lastly, unit tests still need to be implemented.

j-ka11 committed:
Cleaned up some Woocommerce selectors and found a solution to some Woocommerce elements not being on the page at the page's first load.

Before this commit we were not selecting some Woocommerce elements correctly because we were using the sibling operator (~) when it was not needed. Instead, I changed the body selector to be more specific so that we are sure the first selector is selecting the page's body element. The other problem was some of Woocommerce's elements were not on the page immediately on page load, but were added some time in the future after an Ajax request. On most of these requests, Woocommerce triggers actions on the page's body element, so by hooking into those events I was able to add an event listener after the page was loaded. Still need to implement the EventList classes for the rest of the plugins and add unit tests.

j-ka11 committed:
Created WPFormsList class

WPFormsList contains all the necessary info to support the WPForms plugin. Also changed the way javascript is injected into the page in the MeasurementEvent class to be more readable. Also used a for loop to iterate through each node in the javascript query which solves two problems. The first is now we don't try to add event listeners to elements that aren't on the current page. The other is that if there are multiple elements that we want to track for the same event e.g. form-submit, we now add event listeners to each element.

j-ka11 committed:
Updated measurement event list instantiation architecture.

Previously, each subclass of the MeasurementEventList class would store all the event data and then use that data to create a list of MeasurementEvent objects. That method of creation was redundant because that information is already stored in each event, and it also creates a highly coupled relationship between the MeasurementEventList class and the MeasurementEvent class. Now, each subclass directly creates MeasurementEvent objects and only stores those objects.

j-ka11 committed:
Created CF7EventList class and FormidableFormsEventList class

These two classes add all the necessary info to track events for the Contact Form 7 plugin and the Formidable Forms plugins in a similar fashion to the previous subclasses of the MeasurementEventList class.

j-ka11 committed:
Made a bunch of small changes including making sure each Woocommerce event was captured

Some Woocommerce elements are either loaded late with ajax, or their click handler is overwritten at some point. I through together a couple hacky solutions to address these to make sure all Woocommerce events were captured. Still need to add the code to capture Ninja Form events.

j-ka11 committed:
Created NinjaFormsEventList class.

This class contains all the information for the Ninja Forms plugin. There is a bug that when more than one form is on the page, clicking the first submit button will trigger multiple alerts.

j-ka11 committed:
Moved all javascript to be injected into the MeasurementEvent class.

Previously, there was bits of javascript hacks to be injected all throughout the MeasurementCodeInjector class and MeasurementEvent class. It has all been moved into the MeasurmentEvent class and structured in a way that is able to easily be migrated to a system that supports MeasurmentEvent subclasses by plugin.

j-ka11 committed:
Added comments

Classes included: MeasurementEventFactory, CF7EventList, FormidableFormsEventList, MeasurementEvent, MeasurementEventList, WoocommerceEventList, NinjaFormsEventList, and WPFormsEventList

j-ka11 committed:
Added comments to MeasurementCodeInjector class

j-ka11 committed:
Merged all javascript injections into one injection

Previously, we were injecting a script tag for every event. Now, we create a list of JSON objects out of each MeasurementEvent object that contains all the necessary information we need to have a only a few blocks of javascript on the frontend.

j-ka11 committed:
Deleted old injection code (should have been in previous commit)

j-ka11 committed:
Added comments for the single injection of javascript

j-ka11 committed:
Cleaned up minor comments and code bits to be pull-request ready.

Two bugs remain to be fixed. The first has to do with Ninja Forms. We are able to track when a Ninja Form is submitted a single time, but not any times after that. The other bug has to do with Woocommerce. When the user is on the page where they are viewing the cart contents, we curently track three actions: checkout, remove_from_cart, and update_cart. Within Woocommerce, if one of the latter two actions fires, the document that contains those three elements is refreshed with an Ajax request, so our event listeners for all three elements is lost. We need to find a way to hook into that ajax request so we can reapply our event listeners when the ajax request occurs.

yassinh-g commented on src/measurement-events/WoocommerceEventList.php:
```diff
@@ -0,0 +1,36 @@
+            '.woocommerce-page form.woocommerce-checkout');
+        $onEvents = array('click', 'click', 'click', 'click', 'click', 'click', 'click', 'click', 'click', 'submit');
+        $secondLayerSelectors = array(null, null, null, null, 'document.body', null, null, null, null, null);
+        $secondLayerOn = array(null, null, null, null, 'added_to_cart', null, null, null, null, null);
```
This will be hard to maintain given that the index of the element defines in which object the event is defined.

Same for the other plugins

j-ka11 replied:
I was thinking about this problem as well...do you think we should get this information by parsing data in a file that is grouped by event?

yassinh-g commented on src/MeasurementCodeInjector.php:
```diff
@@ -37,11 +55,90 @@ public function setActivePlugins() {
+        $newEvent['on'] = $measurementEvent->getOnEvent();
+        $newEvent['secondLayerSelector'] = $measurementEvent->getSecondLayerSelector();
+        $newEvent['secondLayerOn'] = $measurementEvent->getSecondLayerOnEvent();
+        array_push($this->eventConfigurations, $newEvent);
```
This conversion might not be needed if we directly convert the object to a JSON
https://www.codebyamir.com/blog/object-to-json-in-php

yassinh-g commented on src/MeasurementCodeInjector.php:
```diff
@@ -37,11 +55,90 @@ public function setActivePlugins() {
+                                break;
+                        }
+                        const configCopy = JSON.parse(JSON.stringify(config));
+                        secondLayerNode.on(config.secondLayerOn, function(){
```
Does not need to be addressed in this PR. Can we explore handling dynamically added elements without relying on an use case specific event to attach the listeners.

yassinh-g commented on src/measurement-events/NinjaFormsEventList.php:
```diff
@@ -0,0 +1,30 @@
+        $secondLayerOn = array('nfFormReady');
+
+        $tempEvents = array();
+        for($i = 0; $i < count($selectors); $i++) {
```
No for loop needed. It's easier to read if the MeasurementEvent creation is inlined

j-ka11 replied:
Can you explain a little more what you mean by this comment?

yassinh-g replied:
In this case the for loop only creates a single element therefore no for loop is needed.

yassinh-g commented on src/measurement-events/MeasurementEvent.php:
```diff
@@ -0,0 +1,106 @@
+     * @param $action - string
+     * @param $selector - string
+     * @param $on - string
+     */
```
Not sure if using a builder pattern would make this more readable

https://stackoverflow.com/questions/10961673/php-builder-pattern-without-inner-classes

j-ka11 committed:
Resolved comment about event conversion to JSON.

The list of event configurations is now a direct conversion to json instead of calling an extra function to do the work.

j-ka11 committed:
Implemented Builder Pattern for creating events.

Event creation is now much more readable and inline.

yassinh-g commented on src/measurement-events/WPFormsEventList.php:
```diff
@@ -0,0 +1,19 @@
+    public function __construct() {
+        $builder = MeasurementEvent::createBuilder('WPForms')->category('engagement');
+        $builder->action('form_submit')->selector('.wpforms-submit-container button');
+        $builder->on('click')->secondLayerSelector(null)->secondLayerOn(null);
```
We don't need to set the null values

j-ka11 committed:
Changed Builder pattern to use associative arrays.

Used a different version of the Builder Pattern to make the code more readable. Also don't have to add null values anymore.
